components/sync15/src/state.rs

```diff
@@ -556,142 +394,133 @@ impl<'client, 'keys> SetupStateMachine<'client, 'keys> {
-    HasMetaGlobal(GlobalState),
-    ResolveMetaGlobal(GlobalState, BsoRecord<MetaGlobalRecord>),
-    NeedsFreshCryptoKeys(GlobalState),
-    Ready(GlobalState),
-    FreshStartRequired(GlobalState),
+    // These "Initial" states are only ever used when starting from scratch.
+    Initial,
+    InitialWithConfig {
+        config: InfoConfiguration,
+    },
+    InitialWithInfo {
+        config: InfoConfiguration,
+        collections: InfoCollections,
+    },
+    InitialWithMetaGlobal {
+        config: InfoConfiguration,
+        collections: InfoCollections,
+        global: MetaGlobalRecord,
+        global_timestamp: ServerTimestamp,
+    },
+    WithPreviousState {
+        old_state: GlobalState,
+    },
+    Ready {
+        state: GlobalState,
+    },
+    FreshStartRequired {
+        config: InfoConfiguration,
+        old_global: Option<MetaGlobalRecord>,
+    },
 }
 
 impl SetupState {
     fn label(&self) -> &'static str {
         match self {
-            InitialWithLiveToken(_) => "InitialWithLiveToken",
-            InitialWithLiveTokenAndConfig(_) => "InitialWithLiveTokenAndConfig",
-            InitialWithLiveTokenAndInfo(_) => "InitialWithLiveTokenAndInfo",
-            NeedsFreshMetaGlobal(_) => "NeedsFreshMetaGlobal",
-            HasMetaGlobal(_) => "HasMetaGlobal",
-            ResolveMetaGlobal(_, _) => "ResolveMetaGlobal",
-            NeedsFreshCryptoKeys(_) => "NeedsFreshCryptoKeys",
-            Ready(_) => "Ready",
-            FreshStartRequired(_) => "FreshStartRequired",
+            Initial { .. } => "Initial",
+            InitialWithConfig { .. } => "InitialWithConfig",
+            InitialWithInfo { .. } => "InitialWithInfo",
+            InitialWithMetaGlobal { .. } => "InitialWithMetaGlobal",
+            Ready { .. } => "Ready",
+            WithPreviousState { .. } => "WithPreviousState",
+            FreshStartRequired { .. } => "FreshStartRequired",
         }
     }
 }
 
-/// Whether we should skip fetching `meta/global` or `crypto/keys` from the
-/// server because our locally cached copy is up-to-date, fetch a fresh copy
-/// from the server, or invalidate our locally cached state, then upload a
-/// fresh record.
-enum FetchAction {
-    Skip,
-    Fetch,
-    InvalidateThenUpload,
-}
-
-impl FetchAction {
-    fn from_modified(
-        local: Option<&ServerTimestamp>,
-        remote: Option<&ServerTimestamp>,
-    ) -> FetchAction {
-        match (local, remote) {
-            (Some(local), Some(remote)) => {
-                if *local >= *remote {
-                    FetchAction::Skip
-                } else {
-                    FetchAction::Fetch
-                }
-            }
-            (Some(_), None) => FetchAction::InvalidateThenUpload,
-            _ => FetchAction::Fetch,
-        }
+/// Whether we should skip fetching an item.
+fn is_fresh(local: ServerTimestamp, collections: &InfoCollections, key: &str) -> bool {
+    match collections.get(key) {
```

Comment:
nit: Something like `collections.get(key).map_or(false, |ts| local >= ts)` would be simpler.